### ISSUE TITLE
xread_jitter bails out with EWOULDBLOCK in init_jitter_entropy_source (rngd_jitter.c)

### ISSUE THREAD

**Issue by dauwep** [nhorman/rng-tools#207]
I'm using rngd on an embedded Linux system running on a (quad-core) Cortex A53. After I upgraded rng-tools from 6.11 to 6.16, I'm getting the following error when starting rngd:
```
[jitter]: xread_jitter requests 16 bytes from pipe
[jitter]: failed read: Resource temporarily unavailable
```

Full debug log:
```
~ # /usr/sbin/rngd -f --debug
Initializing available sources
[hwrng ]: Unable to open file: /dev/hwrng
[hwrng ]: Initialization Failed
[jitter]: JITTER timeout set to 5 sec
[jitter]: JITTER attempting to start 4 threads
[jitter]: JITTER thread on cpu 0 wakes up for refill
[jitter]: CPU Thread 0 is ready
[jitter]: JITTER thread on cpu 1 wakes up for refill
[jitter]: CPU Thread 1 is ready
[jitter]: JITTER thread on cpu 2 wakes up for refill
[jitter]: CPU Thread 2 is ready
[jitter]: JITTER thread on cpu 3 wakes up for refill
[jitter]: CPU Thread 3 is ready
[jitter]: Initializing AES buffer
[jitter]: xread_jitter requests 16 bytes from pipe
[jitter]: failed read: Resource temporarily unavailable
... (error message repeats from this point onwards)
```

Looks like `xread_jitter` fails because the read on the pipe in `rngd_jitter.c` bails out with `EWOULDBLOCK`.

While comparing the 6.11 version of `rngd_jitter.c` with 6.16, I noticed the code that adds the `O_NONBLOCK` option to the read end of the pipe used to be at the end of function `init_jitter_entropy_source`, i.e. _after_ the `xread_jitter` calls, while in the 6.16 version, the option is set _before_ these calls. The following obvious patch works around my problem:
```
diff --git a/rngd_jitter.c b/rngd_jitter.c
index 3ff15e1..2264a4e 100644
--- a/rngd_jitter.c
+++ b/rngd_jitter.c
@@ -521,10 +521,6 @@ int init_jitter_entropy_source(struct rng *ent_src)
        CPU_FREE(cpus);
        cpus = NULL;
 
-       flags = fcntl(pipefds[0], F_GETFL, 0);
-       flags |= O_NONBLOCK;
-       fcntl(pipefds[0], F_SETFL, flags);
-
        if (ent_src->rng_options[JITTER_OPT_USE_AES].int_val) {
                /*
                 * Temporarily disable aes so we don't try to use it during init
@@ -588,6 +584,10 @@ int init_jitter_entropy_source(struct rng *ent_src)
 
        }
 
+       flags = fcntl(pipefds[0], F_GETFL, 0);
+       flags |= O_NONBLOCK;
+       fcntl(pipefds[0], F_SETFL, flags);
+
        message_entsrc(ent_src,LOG_DAEMON|LOG_INFO, "Enabling JITTER rng support\n");
        return 0;
 }
```
I assume something's not quite right on my embedded system so this report is a request for advise, rather than an issue ticket.

**Comment by nhorman** [comment]
That change was made specifically to prevent rngd from stalling for long periods of time during start up on systems that produce low entropy using jitter (ARM7 is especially predictable in load/store times, so jitter delays alot).  Without the above change we potentially block for long periods of time on the pipe, above and beyond the timeout value.

So the warnings are normal for low entropy systems, and occasionally lead to failures to init (which is ok on systems that don't produce enough entropy on jitter).

So the short answer to your query is that there isn't actaully anything 'wrong' with your systems, its just that armv7 cpus don't produce alot of entropy with the jitter source, and rngd is telling you that.  It may be that, in your environment, you may need to disable the jitter source at runtime (or compile time if you prefer), and find a different entropy source to use.  If your core supports the rndr instruction, rngd has an entropy source for that which would work much better.  We could also look at implementing a cprng if you like

**Comment by dauwep** [comment]
Thanks for the prompt answer, Neil. As far as I know Cortex A53 is ARMv8 but I do indeed observe a ~50 second stall during startup when I run with the patch mentioned above.

I need to mention I'm using buildroot to build my embedded system (Linux kernel + root file system); the rng-tools 6.11 to 6.16 upgrade fits in the larger context of a buildroot upgrade exercise.

Interestingly, when I downstep rng-tools back to 6.11 in my new buildroot environment, I'm observing approximately the same stalling time. However, when I do a plain copy of the 6.11 executable produced in the baseline buildroot environment, then it works as before: no (observable) stalling.

The only significant difference between the former and latter 6.11 executable is the libjitterentropy version: this went from 2.2.0 to 3.4.1 so that confirms your statement that the jitter entropy source doesn't seem to be the best choice for my system. I'm guessing the new jitterentropy library expects more entropy before it's "satisfied" and returns to a client.

I'll try and look into your suggestion to rely on alternative entropy sources.

**Comment by dauwep** [comment]
Update (and closing comment): we solved this by means of two patches:
1. For the jitterentropy-library, we noticed jitterentropy-base.c is compiled without optimization. The rationale for that is undertandable but this change is reponsible for the bigger part of the large stalling time we observed during startup of our system, after updating to version 3.4.1.
2. The O_NONBLOCK patch in rngd_jitter.c, mentioned higher is still required. Even after reverting to -O2 compilation for jitterentropy-base.c, the system still isn't coming up due to repetitive "failed read: Resource temporarily unavailable", after which it eventually gives up. In our system, this is considered a fatal error.

I understand the above will not be accepted upstream but it might be of some help for users who have older system that don't feature a hwrng.
